From 0f29345103e8ce91a30190e19dd189e9ffa23076 Mon Sep 17 00:00:00 2001
From: Chih-Wei Huang <cwhuang@linux.org.tw>
Date: Mon, 20 Nov 2017 16:50:20 +0800
Subject: [PATCH 12/26] init: handle ctrl-alt-del event

By setting /proc/sys/kernel/ctrl-alt-del to 1, kernel will trap
ctrl-alt-del event and send a SIGINT signal to init.

Start the ctrl-alt-del service on receiving the signal.


Former-commit-id: 6d4d69400fa49fec68993d0e01b21f5371a9d13f
Former-commit-id: b6fd959d7f6d434126063d20ae223e7d4fbb252b
---
 init/reboot_utils.cpp | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/init/reboot_utils.cpp b/init/reboot_utils.cpp
index e6b868e79..faaacc92e 100644
--- a/init/reboot_utils.cpp
+++ b/init/reboot_utils.cpp
@@ -32,6 +32,7 @@
 #include "capabilities.h"
 #include "reboot_utils.h"
 #include "util.h"
+#include "init.h"
 
 namespace android {
 namespace init {
@@ -209,6 +210,12 @@ void InstallRebootSignalHandlers() {
 #endif
     sigaction(SIGSYS, &action, nullptr);
     sigaction(SIGTRAP, &action, nullptr);
+
+    action.sa_handler = [](int sig) {
+        LOG(INFO) << "Got ctrl-alt-del: " << sig;
+        QueueControlMessage("start", "ctrl-alt-del", getpid(), -1);
+    };
+    sigaction(SIGINT, &action, nullptr);
 }
 
 }  // namespace init
-- 
2.34.1

